Hackathon: Check Compatibility Matrix

smk4664 opened the pull request:
<!--
    Thank you for your interest in contributing to Nautobot Dev Example App! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->

## What's Changed

<!--
    Please include:
    - A summary of the proposed changes
    - A sectioned breakdown for larger features under ## subheadings
    - Screenshots, example payloads where relevant:
      - Before/After for bugfixes
      - Using a new feature
-->

This adds a checker for the Compatibility Matrix docs. It tries to determine the minimum/maximum version of Nautobot from the Pyproject.toml (as the min/max version in the __init__.py could be going away). If you provide the --fix flag, it will add or update the line for the compatibility. This only checks the current version, it can't go back in time and check others. This also assumes that the last line is the version, which is the example, but may not be true in every repo.

EDIT:
I purposefully parse the pyproject.toml and compatibility_matrix.md with read instead of using a library like toml so that no additional packages need to be installed in the invoke library.

## To Do

<!--
    Please feel free to update todos to keep track of your own notes for WIP PRs.
-->
- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

smk4664 committed:
Hackathon: Check Compatibility Matrix

This adds a checker for the Compatibility Matrix docs. It tries to determine the minimum/maximum version of Nautobot from the Pyproject.toml (as the min/max version in the __init__.py could be going away). If you provide the --fix flag, it will add or update the line for the compatibility. This only checks the current version, it can't go back in time and check others. This also assumes that the last line is the version, which is the example, but may not be true in every repo.

smk4664 committed:
Create 88.housekeeping

smk4664 committed:
Ruff

gsnider2195 commented on tasks.py:
```diff
@@ -985,3 +987,223 @@ def validate_app_config(context):
+
+def parse_poetry_version_constraint(constraint):
+    """Parse a Poetry version constraint and return (min_version, max_version) as strings."""
+    import re
```
`re` is in the standard library. It shouldn't hurt to have this import at the top

smk4664 replied:
It's already imported, I removed this line.

gsnider2195 commented on docs/dev/release_checklist.md:
```diff
@@ -32,6 +32,7 @@ Every minor version release should refresh `poetry.lock`, so that it lists the m
 
 If there are any changes to the compatibility matrix (such as a bump in the minimum supported Nautobot version), update it accordingly.
+This can now be done using the `inv check-compatibility-matrix --fix`.
 
```
```suggestion

This can be updated automatically using `invoke check-compatibility-matrix --fix`.

```

gsnider2195 commented on changes/88.housekeeping:
```diff
@@ -0,0 +1 @@
+Added a checker for the Compatibility Matrix.
```
```suggestion
Added a new invoke command to check the accuracy of the Compatibility Matrix documentation file and optionally fix it automatically.
```

gsnider2195 commented on .github/workflows/ci.yml:
```diff
@@ -94,6 +94,19 @@ jobs:
           poetry-version: "1.8.5"
       - name: "Linting: markdownlint"
         run: "poetry run invoke markdownlint"
+  check-compatibiility-matrix:
```
Should `check-in-docker` rely on this passing or do we want it to be informational for now?

```suggestion
  check-compatibility-matrix:
```

gsnider2195 reviewed (commented):
I haven't had a chance to review the code yet. Just some initial comments.

smk4664 committed:
Update .github/workflows/ci.yml

Co-authored-by: Gary Snider <75227981+gsnider2195@users.noreply.github.com>

smk4664 committed:
Update changes/88.housekeeping

Co-authored-by: Gary Snider <75227981+gsnider2195@users.noreply.github.com>

smk4664 committed:
Update docs/dev/release_checklist.md

Co-authored-by: Gary Snider <75227981+gsnider2195@users.noreply.github.com>

smk4664 committed:
Remove import, already imported re.

smk4664 committed:
Check-in-docker should rely on the new checker to pass

gsnider2195 reviewed (commented):
If you have time at the end of the sprint could you see if this could be trimmed down? This is a lot of code to add to the tasks.py

smk4664 committed:
Refactor to reduce lines of code and simplify

smk4664 committed:
Ruff

gsnider2195 commented on tasks.py:
```diff
@@ -985,3 +987,203 @@ def validate_app_config(context):
+        return min_v, max_v
+    # Exact version or ==<version>
+    # e.g. 2.0.3 or ==2.0.3
+    m = re.match(r"==?\s*([0-9.]+)", constraint)
```
```suggestion
    m = re.match(r"(?:==)?\s*([0-9.]+)", constraint)
```

The `?` is only making the previous `=` optional but I think you want both of them to be optional?

smk4664 replied:
Yep, good catch. I was trying to not create a grouping, but missed making them both optional.

smk4664 committed:
Update tasks.py

Co-authored-by: Gary Snider <75227981+gsnider2195@users.noreply.github.com>

smk4664 committed:
Merge develop

smk4664 committed:
Simplify the code, remove some of the supported version strings

Nautobot has standardized on a set of the Poetry approved version constraint strings. We only check on this set.

gsnider2195 commented on .github/workflows/ci.yml:
```diff
@@ -153,13 +153,27 @@ jobs:
       - name: "Linting: markdownlint"
         run: "poetry run invoke markdownlint"
+  check-compatibility-matrix:
+    runs-on: "ubuntu-22.04"
```
What year is this? 😄 
```suggestion
    runs-on: "ubuntu-latest"
```

Defiantearth commented on .github/workflows/ci.yml:
```diff
@@ -153,13 +153,27 @@ jobs:
+      - name: "Setup environment"
+        uses: "networktocode/gh-action-setup-poetry-environment@v6"
+        with:
+          poetry-version: "1.8.5"
```
```suggestion
          poetry-version: "2.1.3"
```

Defiantearth commented on tasks.py:
```diff
@@ -1082,3 +1084,101 @@ def validate_app_config(context):
+    nautobot_min, nautobot_max = parse_poetry_version_constraint(nautobot_match[1] or nautobot_match[2])
+
+    lines = matrix_path.read_text().splitlines(keepends=True)
+    last_index = next((i for i in range(len(lines) - 1, -1, -1) if lines[i].startswith("| ")), None)
```
This won't work for something like [nautobot_tools](https://github.com/networktocode-llc/nautobot-app-tools/blob/develop/docs/admin/compatibility_matrix.md?plain=1).

Defiantearth commented on .github/workflows/ci.yml:
```diff
@@ -153,13 +153,27 @@ jobs:
+        with:
+          poetry-version: "1.8.5"
+      - name: "Linting: check-compatibility-matrix"
+        run: "poetry run invoke check-compatibility-matrix"
```
Maybe instead of a CI workflow we would prefer the release workflow to just run the command with --fix when running. I'm worried that that could fail too often though. Maybe the CI is the correct choice. We also should discuss in standup today if patch releases are okay for nautobot min bumps if so we should handle those too.

Defiantearth reviewed (commented):
Looks Great so far just a few comments.
